Worker Configurations

jhford opened the pull request:
This pull request includes worker-configurations and standard error infrastructure for the project.  The actual format which worker configurations needs to resolve down to is yet to be determined, but has a bug on file: bug 1487671.

jhford committed:
Fix small bug in Array handling for deep assignments

jhford committed:
Bug 1487668 - [worker-manager] Implement Worker Configuration

jhford committed:
Bug 1497688 - Implement standardised error handling

jhford committed:
Convert code to use standard errors

jhford committed:
use matching node versions in package.json and .travis.yml

imbstack commented on lib/errors.js:
```diff
@@ -0,0 +1,50 @@
+// as a standardised method of throwing these errors
+
+// These are the strings for which we want to generate matching error types.
+// They will also be set as the code property on the instances of each error
```
👍

imbstack commented on lib/worker-config.js:
```diff
@@ -0,0 +1,293 @@
+ *
+ * The constructor parameters are:
+ *  - id: identifier for this worker configuration
+ *  - workerTypes: list of objects {workerType, biddingStrategy, providers}
```
I guess this has changed names to `workerTypeConfigurations`?

imbstack commented on lib/worker-config.js:
```diff
@@ -0,0 +1,293 @@
+      if (typeof workerType !== 'string') {
+        this._throw(errors.InvalidWorkerConfiguration, 'worker type name must be string');
+      }
+      this.workerType = workerType;
```
I'm a bit confused by this. Wouldn't we overwrite the `workerType` on every loop. Was this an `Array.forEach()` before?

jhford replied:
hmm, it was always a for-of, but I think it's a typo.  I've removed the assignments to `this` in this loop in my current patch set.

jhford replied:
Ahh, I remember, this code was originally in a `.map` in the factory function, where assignment was important, but I decided to move the validation into the constructor so that both factory-built instances had the same validation as those built with the constructor.

imbstack commented on lib/worker-config.js:
```diff
@@ -0,0 +1,293 @@
+
+  /**
+   * Evaluate a WorkerConfiguration's rules based on a set of satisfiers and a
+   * map of providers.  The satisfiers must contain a provisionerId which is an
```
Why not `providerId` if it is a provider?

jhford replied:
The providers would be passed in as a Map but the key from that providers would be looked up by the value in the satifiers.  The thinking is that we want to specify the id in as few places as possible.

After submitting this PR, I started working on a script to do conversions from AWS Provisioner worker types into rule sets and realised that there's a couple important changes we're going to need to make to worker configurations, one of which removes the providers reference from evaluate.

I think it's probably better to push the new changes than try to explain them in this thread?  The result is a lot cleaner and I think having the example of a realistic Worker Configuration makes this a lot more clear.  If I don't hear anything, I'll push to this branch :)

imbstack commented on lib/worker-config.js:
```diff
@@ -0,0 +1,293 @@
+ * configuration.  The configuration must be equal to the result of evaluating
+ * a successful WorkerConfiguration.evaluate().
+ */
+class StaticWorkerConfiguration extends WorkerConfiguration {
```
Can the docstring for this describe why it might be used vs a provisioned one?

jhford replied:
Sure!  Though, that said, the changes that I've been making might actually make the entire class superfluous.

imbstack commented on lib/worker-config.js:
```diff
@@ -0,0 +1,293 @@
+
+    let requiredSatisfiers = this.rules.requiredSatisfiers().sort();
+    let providedSatisfiers = provider.providedSatisfiers();
+    if (!Array.isArray(providedSatisfiers)) {
```
Isn't this asserted when the provider is defined?

jhford replied:
I've gotten rid of this assertion in the code, but you're right that it would've been unnecessary once we had a real Fake Provider (:laughing:) to test against, but my concern is that the fake provider in my tests were sort of... simple, so I wanted to be overzealous up front.

imbstack commented on lib/worker-config.js:
```diff
@@ -0,0 +1,293 @@
+  }
+
+  if (config.rules && config.configuration) {
+    throw new Error('Worker Configuration must be static or provisioned');
```
Let's add specifically what is wrong with the config here. Otherwise I see myself digging into the code to see exactly what I did wrong.

jhford replied:
I agree.  I'm going to reevaluate the distinction between static and provisioined worker configurations in general, but I just couldn't think of a good error message.

imbstack reviewed (changes requested):
Just some nits and one refactor-typo. Looks good otherwise!

jhford commented:
@imbstack Hey, thanks for the review!  I've addressed your comments in my local branch.  When working on the next bit of work, I realised there was a pretty minor change to `Ruleset`s which would have a pretty major impact for the whole worker configuration system.

The change is to remove the requirement of having every possible satisfier specified when evaluating conditions.  Instead of throwing, it'll evaluate the condition to be 'unmet'. 

Take for example the conditions definition:
```javascript
conditions: {
  a: 'a',
}
```
Before passing a satisfiers of `{b: 'john'}` would throw an error, but with the change would evaluate to `{}`.

I think rather than explain everything in my head on this, it is probably better to push to the branch?  I won't edit these commits, so it shouldn't be hard to see only the new stuff.

jhford committed:
Bug 1487671 - [worker-manager] Implement internal configuration for worker types

jhford committed:
Bug 1498264 - [worker-manager] implement functionality to convert aws-provisioner worker type definition to a worker configuration

jhford commented:
@imbstack I think I addressed all of the concerns, some by removing.  I also included the current worker type to new worker configuration in this PR to avoid having a seperate PR that's mostly related.  Please let me know if you'd like me to break it up any further.  I didn't change any of the already reviewed commits, in case you'd prefer to look at the changes instead of the complete new version.

imbstack reviewed (approved):
Sweet. The refactor really made it nice.

jhford commented:
thanks :)
